drivers/regulator/regulator_da1469x.c

```diff
@@ -0,0 +1,451 @@
+	case VDD_CLAMP:
+		voltage_ix = POWER_REG_GET(VDD_CLAMP_LEVEL);
+		break;
+	case VDD:
+		voltage_ix = POWER_REG_GET(VDD_LEVEL);
+		break;
+	case VDD_SLEEP:
+		voltage_ix = POWER_REG_GET(VDD_SLEEP_LEVEL);
+		break;
+	case V18:
+		voltage_ix = POWER_REG_GET(V18_LEVEL);
+		break;
+	case V14:
+		voltage_ix = POWER_REG_GET(V14_LEVEL);
+		break;
+	case V30:
+		voltage_ix = POWER_REG_GET(V30_LEVEL) >> 1;
+		break;
+	default:
+		voltage_ix = 0;
+		break;
+	}
+	*volt_uv = config->voltages[voltage_ix];
+
+	return 0;
+}
+
+static int regulator_da1469x_set_current_limit(const struct device *dev,
+					       int32_t min_ua, int32_t max_ua)
+{
+	const struct regulator_da1469x_config *config = dev->config;
+	int ret = 0;
+
+	switch (config->rail) {
+	case VDD:
+		DCDC_REG_SET(VDD, CUR_LIM_MAX_HV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(VDD, CUR_LIM_MAX_LV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(VDD, CUR_LIM_MIN, ((max_ua / 30000) - 1));
+		break;
+	case V14:
+		DCDC_REG_SET(V14, CUR_LIM_MAX_HV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(V14, CUR_LIM_MAX_LV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(V14, CUR_LIM_MIN, ((max_ua / 30000) - 1));
+		break;
+	case V18:
+		DCDC_REG_SET(V18, CUR_LIM_MAX_HV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(V18, CUR_LIM_MAX_LV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(V18, CUR_LIM_MIN, ((max_ua / 30000) - 1));
+		break;
+	case V18P:
+		DCDC_REG_SET(V18P, CUR_LIM_MAX_HV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(V18P, CUR_LIM_MAX_LV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(V18P, CUR_LIM_MIN, ((max_ua / 30000) - 1));
+		break;
+	default:
+		ret = -ENOSYS;
+		break;
+	}
+
+	return ret;
+}
+
+static int regulator_da1469x_get_current_limit(const struct device *dev,
+					       int32_t *curr_ua)
+{
+	const struct regulator_da1469x_config *config = dev->config;
+	int ret = 0;
+
+	switch (config->rail) {
+	case VDD:
+		*curr_ua = (DCDC_REG_GET(VDD, CUR_LIM_MAX_HV) - 1) * 30000;
+		break;
+	case V14:
+		*curr_ua = (DCDC_REG_GET(V14, CUR_LIM_MAX_HV) - 1) * 30000;
+		break;
+	case V18:
+		*curr_ua = (DCDC_REG_GET(V18, CUR_LIM_MAX_HV) - 1) * 30000;
+		break;
+	case V18P:
+		*curr_ua = (DCDC_REG_GET(V18P, CUR_LIM_MAX_HV) - 1) * 30000;
+		break;
+	default:
+		ret = -ENOSYS;
+	}
+
+	return ret;
+}
+
+static const struct regulator_driver_api regulator_da1469x_api = {
+	.enable = regulator_da1469x_enable,
+	.disable = regulator_da1469x_disable,
+	.count_voltages = regulator_da1469x_count_voltages,
+	.list_voltage = regulator_da1469x_list_voltage,
+	.set_voltage = regulator_da1469x_set_voltage,
+	.get_voltage = regulator_da1469x_get_voltage,
+	.set_current_limit = regulator_da1469x_set_current_limit,
+	.get_current_limit = regulator_da1469x_get_current_limit,
+};
+
+static int regulator_da1469x_init(const struct device *dev)
+{
+	const struct regulator_da1469x_config *config = dev->config;
+
+	regulator_common_data_init(dev);
+
+	if (config->rail == V30) {
+		POWER_REG_SET(LDO_3V0_REF, DT_INST_PROP_OR(DT_NODELABEL(V30),
```

Comment:
this is unsafe: what if we have >1 instance of V30? please, store the value in driver's config

Reply:
There is no point is storing it drivers config. There is only one V30 rail in this soc.

Reply:
yes, but driver is designed to be multi-instance, and this construct is wrong.

Reply:
valued set from config